### ISSUE TITLE
Rule Engine rules and Performance

### ISSUE THREAD

**Issue by jainh** [jdereg/n-cube#210]
Hi, 
Can n-cube handle following points:

1. more complex rule statements (language), and relationships among rules such as inclusive, exclusive rules (AND, XOR), nested rules etc.
2. Conflict resolution (show where rules conflict with one another) and rule precedence. When multiple rules are in conflict, which one takes effect.
3. Blacklist/whitelist support.
4. Rule application beyond just user, and org (establish different tiers), tools etc., different types of resources; management of resources;

Is there any performance benchmarking done on this project ?

Thank you.

**Comment by jdereg** [comment]
#1:
N-Cube is Turing complete when it comes to rules.  It maintains a linear list of rules that are run in order.  It does not use the RETE algorithm.  This makes it easy to follow.  The condition for a rule is any valid Groovy statement, or you can use a URL to a Groovy class with a run() method that will have the current input, output maps passed to it.  This approach allows for rules with condition as simple as a hard-coded true, false, as well as input.age > 10, etc.  Keeps the knowledge of package, imports, class, and method declaration out of the logic making it easier for the business user to read.

If the condition is true, the associated statement is executed.  Again, same thing for the statement cell.  It can be as simple as a primitive value, or a groovy expression, or a URL to a Groovy class.

Rule axis can be combined with additional dimensions (decision variables) if desired.  A rule axis can be set to fire the first encountered true condition.  Or it can be set to fire all true conditions.

Rule axes support a 'default' column, which if present is executed if NO conditions have fired.

#2 
Rule order is chosen by the rule author like is done with imperative code.  The condition listed first is evaluated and optionally fired, the second condition is evaludated, and optionally fired, and so on until the rule list is exhausted.  In addition, as mentioned above, if the rule axis is set to  fire once, then it will stop executing rules after the first condition evaluates to true.

#3 
Blacklist versus whitelist support.  The default column combined with your rules is how blacklist versus white list is done.  You can list the exceptions as individual rules, and then have the Default rule (last condition) handle the common case.  Or you can list all the rules, and have the Default handle the exception case.  The way in which you choose to do this depends on whether a black list or white list approach is needed.

The Default column on a rule axis is the NOT (!) of each condition ANDed together.  The !(cond1 && cond2 && ... cond-n).

#4 
Not exactly what your question is here?  The rule condition and statements have access to everything in the classpath (which is dynamically augmented by the addition list of classpath roots in the sys.classpath n-cube).  An infinite number of condition variables are available to the condition and statement.

The decision variable values are placed on an input Map instance.  When the ncube.getCell(input, output) is called, each condition has access to all of the inputs values.  Furthermore, the condition can access other n-cubes (rule or non-rule decision trees supporting primary conditions, secondary conditions, tertiary conditions, and so on).  Furthermore the statements can access other n-cubes (rule or non-rule).

#5 (Benchmarking question)
Internally when built there are many performance tests that are run to ensure that the library is running fast and does not degrade in performance when changes are made.

Because n-cube compiles the condition and statement source into Classes that are dynamcially added to the JVM, the condition and statement code run as fast as your normal compiled code.

And because Groovy is used, the syntax of the rule conditions is much clean than typcial code - no semi colons, optional parenthesis, full access to JDK, network access, and so on.  New APIs can be added that are then easily accessible by the rule conditions and statements.

If the rule condition source or statement source is changed, the code cache can be cleared live, and n-cube will detect that the associated class is missing, and it will recompile the rule conditions and statements dynamically and continue on, without interruption.  This allows development of the business logic to stay outside of the deployed .war file, so changes can be quickly and easily tested.

**Comment by jainh** [comment]
@jdereg 
Thank you for prompt and detailed explanation. 
A few more questions 
1) Does it support hierarchical rules ?
2) Do you have documentation of Api ?
3) It support Decision tables, trees and template for rules, is that right ?

Thank you

**Comment by jdereg** [comment]
#1 
Yes NCUBE supports hierarchical rules.  In this case, the statement that executes for the condition joins to an n-cube with the secondary rule set, and so on (as deep as needed).

#2 
The entire API is documented and the documentation is something that is constantly being improved, just like the code.  See http://www.javadoc.io/doc/com.cedarsoftware/n-cube for the entire NCUBE API docs.  

The main classes you interact with at a programmatic level, is NCube, NCubeManager.  As you use NCUBE more and more, you will want to become familiar with Axis and Column.

There is a completely developed user interface that allows NCUBEs to be visualized, edited, and executed with all the enterprise-class features like complete revision history for each n-cube, tenant and application separation of n-cubes, versioning, release management, and branch management (multiple users working on a set of n-cubes in their own branches and updating / committing them).  See it at http://ncube.io.  This version is a little bit older and will be updated soon (tied in so that it is automatically updated as the code is committed).

#3 
NCUBE supports decision tables, trees, and templating (much like velocity using GroovyTemplate's for implementation).  

The business domain areas useful for n-cube are Pricing, Rating, Commission management, and any other areas where there a large sets of supporting data tables (like insurance ISO tables) along with complex rules (methodologies) that interact with these tables.

The N-Cube-Editor (NCE) is written as a Business IDE.  The intended user is a business analysis or developer familiar with the business problem domain area, but not necessarily a hard-code developer-type.   This approach allows Business Analysts at companies to shift from writing Documents to actual making changes to the actual system.

In a nutshell, systems architected on NCUBE expose more 'surface area' to the business user, so that they can take more control over their systems.  In the past, most of this type of data and rules ends up hard-coded in COBOL, Java, etc. and the best the Business Analysts can do is write about what they want changed, and the wait (and hope) for it to be implemented (with usually a long and painful wait).
